contrib/renew-client-cert

```diff
@@ -0,0 +1,66 @@
+#!/bin/sh -e
+
+# This is an OnChanged handler that can handle requests to renew the client
+# certificates used by the device to connect to the device gateway
+
+[ -z "$CONFIG_FILE" ] && (echo "No CONFIG_FILE specified"; exit 1)
+[ -f "$CONFIG_FILE" ] || (echo "$CONFIG_FILE does not exist"; exit 1)
+[ -z "$SOTA_DIR" ] && (echo "No SOTA_DIR specified"; exit 1)
+[ -d "$SOTA_DIR" ] || (echo "$SOTA_DIR does not exist"; exit 1)
+
+fioconfig=$(which fioconfig)
+
+timestamp=$(cat $CONFIG_FILE | grep TIMESTAMP | cut -d= -f2)
+[ -z "$timestamp" ] && (echo "No TIMESTAMP found in config file"; exit 1)
+echo "Timestamp for rotation is $timestamp"
+
+est_server=$(cat $CONFIG_FILE | grep ESTSERVER | cut -d= -f2)
+[ -z "$est_server" ] && (echo "No ESTSERVER found in config file"; exit 1)
+echo "EST server for rotation is $est_server"
+
+# We need to toggle between two different pkcs11 slot ids to rotate on HSMs:
+pkey_ids=$(cat $CONFIG_FILE | grep PKEYIDS | cut -d= -f2)
+if [ -n "$pkey_ids" ]; then
+	echo "PKCS11 key IDs: ${pkey_ids}"
+fi
+cert_ids=$(cat $CONFIG_FILE | grep CERTIDS | cut -d= -f2)
+if [ -n "$cert_ids" ]; then
+	echo "PKCS11 cert IDs: ${cert_ids}"
+fi
+
+rotate() {
+	extra_args=""
+	if [ -n "$pkey_ids" ]; then
+		extra_args="--pkcs11-key-ids=$pkey_ids"
+	fi
+	if [ -n "$cert_ids" ]; then
+		extra_args="$extra_args --pkcs11-cert-ids=$cert_ids"
+	fi
+	exec $fioconfig -c $SOTA_DIR renew-cert $est_server "$extra_args"
+}
+
+# We'll execute under a few conditions:
+# 1) A cert rotation has never taken place (cert-rotation.state.completed does not exist)
+# 2) We need to do a new rotation ($timestamp > timestamp of cert-rotation.state.completed)
+# 3) A rotation seems stuck (cert-rotation.state exists)
+
+active_file="${SOTA_DIR}/cert-rotation.state"
+completed_file="${active_file}.completed"
+
+if [ -f "$active_file" ] ; then
+	echo "Rotating certs: cert-rotation.state exists"
+	rotate
+fi
+
+if [ ! -f "$completed_file" ] ; then
+	echo "Rotating certs: cert-rotation.state.completed does not exist"
+	rotate
+fi
+
+last_ts=$(stat -c %Z "$completed_file")
+if [ $last_ts -lt $timestamp ] ; then
```

Comment:
Where `$TIMESTAMP` is originated?

If `$timestamp`/`$TIMESTAMP` is in the future compared to a device local time then  the rotation will happen again.
Maybe it makes to check the $TIMESTAMP value if it's not originated on a device.

Reply:
I've force pushed a new revision that uses a "Rotation ID" instead of timestamp. I think this will work better than a timestamp. Then fioctl can just use something like a uuid to set this field. Let me know what you think

Reply:
I think it should work.

It even can help with user usability of this feature if we store the ongoing cert rotation ID in the DB, so a user and fioctl can check if there is one and inform about its status.

Reply:
yep. and we keep the last 10 config versions, so we will have some info and could make that better w/o much effort